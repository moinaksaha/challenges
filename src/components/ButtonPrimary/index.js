// @flow
import React, { Component } from 'react'
import styled from 'styled-components';

const ButtonWrapper = styled.div`
  padding: 5px 10px;
  border: 2px solid #648cf3;
  color: #648cf3;
  font-weight: bold;
  font-size: 14px;
  border-radius: 4px;
  cursor: pointer;
  transition: 0.3s;

  :hover{
      background: #648cf3;
      color: #ffffff;
  }
`;

class ButtonPrimary extends Component {
  render() {
    return (
      <ButtonWrapper>
        {this.props.displayText}
      </ButtonWrapper>
    );
  }
};

export default ButtonPrimary


