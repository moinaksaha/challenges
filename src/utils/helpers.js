export const summaryDonations = (danations) => (
  danations.reduce((accumulator, value) => (accumulator + value))
);

export const sumIndividualDonations = (donations) => {
  if (donations && donations.length) {
    const sum = donations.reduce((acc, item) => (
      {
        amount: acc.amount + item.amount,
        breakdown: `${(acc.breakdown !== '')? `${acc.breakdown}, ${item.amount}`: `${item.amount}`}`, //${acc.breakdown}, ${item.amount}`,
      }
    ), {
      amount: 0,
      breakdown: '',
    })
    return sum;
  }
  return null;
}