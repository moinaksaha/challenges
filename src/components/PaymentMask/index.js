// @flow

// Necessary Imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';

// Styled components 
// Reference - https://github.com/styled-components/styled-components
import styled, { keyframes } from 'styled-components';

// ActionCreators imports
import { makePayment, clearPaymentInProgressData } from '../../store/actions/donations';

// import loadable from '@loadable/component';
// const loaderImage = loadable(() => import('../../images/loader.gif'));
// const ButtonPrimary = loadable(() => import('../ButtonPrimary/index.js'));
// const GenericMessage = loadable(() => import('../GenericMessage/index.js'));

// Importing other components
import ButtonPrimary from '../ButtonPrimary';
import GenericMessage from '../GenericMessage';
import loaderImage from '../../images/loader.gif';

// Using Keyframes for animation
const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

// Styled Components
const Mask = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(255, 255, 255, 0.96);
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  animation: 0.3s ${fadeIn} ease-in;

  & >p{
    font-size: 16px;
    color: #5c606f;
  }

  & >div{
    padding: 5px;
  }
`;

// Styled Components
const PaymentOptionHolder = styled.div`
  display: flex;

  & >label{
    display: flex;
    padding: 5px;
    color: #5c606f;
  }
`;

// Styled Components
const ClosePaymentButton = styled.div`
  position: absolute;
  top: 20px;
  right: 25px;
  font-size: 12px;
  font-weight: bold;
  cursor: pointer;
`;

// Styled Components
const PaymentHelpers = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: 13px;
`;

// Styled Components
const ErrorHolder = styled.div`
  font-size: 12px;
  position: absolute;
  bottom: 10px;
  margin: 0px auto;
  animation: 0.3s ${fadeIn} ease-in;
`;

class PaymentMask extends Component {

  constructor(props) {
    super(props);
    // Initial state
    this.state = {
      selectedAmount: null,
      paymentInProgress: false,
      showError: false,
    }
  }

  // Resets payment data when the paymentStatus.loading prop changes from true to false
  componentDidUpdate = prevProps => {
    if (prevProps.paymentStatus.loading === true && this.props.paymentStatus.loading === false) {
      setTimeout(() => {
        this.clearPaymentData();
      }, 2000);
    }
  }

  // Function to trigger close and reset of payment data
  clearPaymentData = () => {
    this.handleCloseButton();
    const { clearPaymentInProgressData } = this.props;
    clearPaymentInProgressData();
    this.resetPaymentForm();
  }

  // Actual function to clear form data
  resetPaymentForm = () => {
    this.setState({
      selectedAmount: null,
      paymentInProgress: false,
      showError: false,
    })
  }

  // Function to handle close button click
  handleCloseButton = () => {
    this.resetPaymentForm();
    const { handleCloseButtonClick } = this.props;
    handleCloseButtonClick();
  }

  // Function to initiate Payment API call if payment amount is selected
  // if nothing is selected - show appropriate error message
  initiatePayment = () => {
    const { makePayment, data } = this.props;
    const { selectedAmount } = this.state;
    const uuidv1 = require('uuid/v1');
    if (selectedAmount) {
      this.setState({
        paymentInProgress: true,
      })
      makePayment({
        charitiesId: data.id, 
        amount: selectedAmount, 
        currency: data.currency,
        id: uuidv1,
      })
    } else {
      this.setState({
        showError: true,
      })
    }
  }

  // Update the selected payment amount on radio button selection
  handlePaymentValueChange = (event) => {
    this.setState({
      selectedAmount: parseInt(event.target.value),
      showError: false,
    })
  }

  render() {
    const { data, visible, paymentStatus } = this.props;
    const { paymentInProgress, selectedAmount, showError } = this.state;

    // return null if visible is false
    if (!visible) {
      return null;
    }

    // map through the available payments options for donation
    const payments = [10, 20, 50, 100, 500].map((amount, j) => (
      <label key={j}>
        <input
          type="radio"
          name="payment"
          value={amount}
          checked={this.state.selectedAmount === amount}
          onChange={this.handlePaymentValueChange}
        /> {amount}
      </label>
    ));

    // Only show this section when payment is in progress
    // hides the payment options screen
    if (paymentInProgress) {
      // show loading if loading status is true
      if (paymentStatus.loading) {
        return (
          <Mask>
            <PaymentHelpers>
              <img src={loaderImage} width={30} height={30} />
              <GenericMessage 
                message={'Processing Payment...'}
                type={'inProgress'}
              />
            </PaymentHelpers>
          </Mask>
        )
      } 
        
      // show error if error status thrown
      if (paymentStatus.error) {
        return (
          <Mask>
            <PaymentHelpers>
              <GenericMessage 
                message={'Something went wrong. Please try again!'}
                type={'error'}
              />
            </PaymentHelpers>
          </Mask>
        )
      }

      // Show thanks message when payment successful
      if (paymentStatus.data) {
        return (
          <Mask>
            <PaymentHelpers>
              <GenericMessage 
                message={`Thanks for donating ${selectedAmount} ${data.currency}`}
                type={'success'}
              />
            </PaymentHelpers>
          </Mask>
        )
      }

      return null;
    }

    // Payment options holder
    return (
      <Mask>
        <ClosePaymentButton onClick={this.handleCloseButton}>X</ClosePaymentButton>
        <p>{`Select the amount to donate (${data.currency})`}</p>
        <PaymentOptionHolder>
          {payments}
        </PaymentOptionHolder>
        <div onClick={this.initiatePayment}>
          <ButtonPrimary displayText={'Pay'} />
        </div>
        {showError && 
            <ErrorHolder>
              <GenericMessage message={'Please select an amount to donate first.'} type={'error'} />
            </ErrorHolder>
        }
      </Mask>
    )
  }
};

// Proptype Validations
PaymentMask.propTypes = {
  makePayment: PropTypes.func.isRequired,
  clearPaymentInProgressData: PropTypes.func.isRequired,
  paymentStatus: PropTypes.object.isRequired,
  data: PropTypes.shape({
    currency: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
  }), 
  visible: PropTypes.bool,
  handleCloseButtonClick: PropTypes.func.isRequired,
};

// default props
PaymentMask.defaultProps = {
  visible: false,
};

const mapStateToProps = state => {
  return {
    paymentStatus: state.donations.paymentStatus,
  };
};
    
const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      makePayment,
      clearPaymentInProgressData,
    },
    dispatch
  );
    
export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(PaymentMask)
);


