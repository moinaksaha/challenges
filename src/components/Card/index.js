// @flow
import React, { Component } from 'react';
import styled from 'styled-components';

import { sumIndividualDonations } from '../../utils/helpers';

import ButtonPrimary from '../ButtonPrimary';
import PaymentMask from '../PaymentMask';

const CardImage = styled.div`
  min-height: 250px;
  background-image: url(${props => props.img});
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center center;
`;

const CardWrapper = styled.div`
  flex: 0 50%;
`;

const Container = styled.div`
  position: relative;
  margin: 20px;
  box-shadow: 0 1px 20px 0 rgba(0,0,0,.3);
  display: flex;
  flex-direction: column;
  border-radius: 6px;
  overflow: hidden;
  transition: 0.5s;

  :hover{
    box-shadow: 0 1px 30px 0 rgba(0,0,0,.5);
  }
`;

const TitleBar = styled.div`
  padding: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const CharityName = styled.div`
  color: #5c606f;
  font-weight: 500;
`;



const PreviousDonationBreakdown = styled.div`
  font-size: 10px;
  padding: 5px 0;
  color: #aaa;
`;

class Card extends Component {

  constructor(props) {
    super(props);
    this.state = {
      showPaymentButton: false,
    }
  }

  showPaymentScreen = () => {
    this.setState({
      showPaymentButton: true,
    })
  }

  hidePaymentScreen = () => {
    this.setState({
      showPaymentButton: false,
    })
  }

  render() {
    const { data, donations } = this.props;
    const { showPaymentButton } = this.state;

    // const payments = [10, 20, 50, 100, 500].map((amount, j) => (
    //   <label key={j}>
    //     <input
    //       type="radio"
    //       name="payment"
    //     /> {amount}
    //   </label>
    // ));

    const previousDonationsSummary = sumIndividualDonations(donations);

    return (
      <CardWrapper>
        <Container>
          <PaymentMask 
            visible={showPaymentButton} 
            data={data}
            handleCloseButtonClick={this.hidePaymentScreen}
          />
          {/* {showPaymentButton && 
            <Mask>
              <ClosePaymentButton onClick={this.hidePaymentScreen}>X</ClosePaymentButton>
              <div>{`Select the amount to donate (${data.currency})`}</div>
              <PaymentOptionHolder>
                {payments}
              </PaymentOptionHolder>
              <div>
                <ButtonPrimary displayText={'Pay'} key={'charitypaybutton'} />
              </div>
            </Mask>
          } */}
          <CardImage img={`../images/${data.image}`}></CardImage>
          <TitleBar>
            <div>
              <CharityName>{`${data.name}`}</CharityName>
              {previousDonationsSummary && 
                <PreviousDonationBreakdown>
                  {`${data.currency} ${previousDonationsSummary.amount}  (${previousDonationsSummary.breakdown})`} 
                </PreviousDonationBreakdown>
              }
              
            </div>
            
            <div onClick={this.showPaymentScreen}>
              <ButtonPrimary displayText={'Donate'} key={'donatebutton'}/>
            </div>
          </TitleBar>
        </Container>
      </CardWrapper>
    );
  }
};

export default Card

