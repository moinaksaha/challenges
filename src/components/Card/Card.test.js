import React from 'react'
import Card from './';
import ButtonPrimary from '../ButtonPrimary';
import PaymentMask from '../PaymentMask';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { mountWrap, shallowWrap } from '../../utils/testUtils';

import sinon from 'sinon';

describe('Component: <Card />', () => {

  test('Snapshot: Enzyme', () => {
    const tree = shallow(<Card />)
    expect(toJson(tree)).toMatchSnapshot()
  })

  test('Render Status: Success', () => {
    const wrapper = shallow(<Card />);
    expect(wrapper.exists()).toBe(true);
  })
  
  const mockData = {
    currency:'THB',
    id:1,
    image:'baan-kru-noi.jpg',
    name: 'Baan Kru Noi',
  }

  const mockDonations = [
    {
      charitiesId: 1,
      amount: 20,
      currency: 'THB',
      id: 2,
    },
  ]

  test('Snapshot[With Mock Data]: Enzyme', () => {
    const wrapper = shallow(<Card data={mockData} donations={mockDonations}/>)
    expect(wrapper).toMatchSnapshot()
  })

  test('Checking Initial state of the component', () => {
    const wrapper = shallow(<Card />);
    expect(wrapper.state().showPaymentButton).toBe(false);
  })

  test('Initial State', () => {
    const wrapper = shallow(<Card />);
    expect(wrapper.state().showPaymentButton).toBe(false);
  })

  test('Testing children: PaymentMask', () => {
    const wrapper = shallow(<Card />);
    expect(wrapper.find(PaymentMask).length).toEqual(1);
  })

  test('Testing children: ButtonPrimary', () => {
    const wrapper = shallow(<Card />);
    expect(wrapper.find(ButtonPrimary).length).toEqual(1);
  })

  test('State changes -> showPaymentButton: true -> on clicking show button', () => {
    const wrapper = shallow(<Card />);
    wrapper.find('#showPaymentScreenButton').simulate('click');
    expect(wrapper.state().showPaymentButton).toBe(true);
  })

  test('Passes new value of visible Prop to PaymentMask component on button click', () => {
    const wrappedMount = () => shallowWrap(<Card />);
    const wrapper = wrappedMount();
    wrapper.find('#showPaymentScreenButton').simulate('click');
    expect(wrapper.find(PaymentMask).prop('visible')).toEqual(wrapper.state().showPaymentButton);
  });

  test('Passes down correct data prop to PaymentMask component', () => {
    const wrappedMount = () => shallowWrap(<Card data={mockData} />);
    const wrapper = wrappedMount();
    expect(wrapper.find(PaymentMask).prop('data')).toEqual(mockData);
  });

  // test()

  // test('calls componentDidMount', () => {
  //   const spy = sinon.spy(Card.prototype, 'componentDidMount');

  //   const wrappedMount = () => shallowWrap(<Card data={mockData} componentDidMount={spy} />);
  //   const wrapper = wrappedMount();

  //   // const wrapper = mount(<Card componentDidMount={spy} />);
  //   expect(spy.calledOnce).to.equal(true);
  //   // Clean up
  //   spy.restore();
  // });

})