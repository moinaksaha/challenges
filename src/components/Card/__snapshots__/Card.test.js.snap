// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Card /> Snapshot: Enzyme 1`] = `
<styled.div>
  <styled.div>
    <withRouter(Connect(PaymentMask))
      data={
        Object {
          "currency": "NA",
          "id": 0,
          "image": undefined,
          "name": "Donate",
        }
      }
      handleCloseButtonClick={[Function]}
      visible={false}
    />
    <styled.div
      img="../images/undefined"
    />
    <styled.div>
      <div>
        <styled.div>
          Donate
        </styled.div>
      </div>
      <div
        id="showPaymentScreenButton"
        onClick={[Function]}
      >
        <ButtonPrimary
          displayText="Donate"
          key="donatebutton"
        />
      </div>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Component: <Card /> Snapshot[With Mock Data]: Enzyme 1`] = `ShallowWrapper {}`;
